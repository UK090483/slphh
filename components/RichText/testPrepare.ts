import { SanityBlock } from "@lib/SanityPageBuilder/lib/RichText";

import { v4 as uuid } from "uuid";

interface GetTestBlockProps extends Partial<SanityBlock> {}

type SpanProps = {
  text: string;
  marks?: string[];
};
export const span = (props: SpanProps) => {
  return {
    _key: uuid(),
    _type: "span",
    marks: props.marks || [],
    text: props.text,
  };
};
export const getTestBlock = (props: GetTestBlockProps = {}) => {
  return {
    _key: uuid(),
    _type: "block",
    children: [span({ text: "testText" })],
    style: "normal",
    markDefs: [],
    ...props,
  } as SanityBlock;
};

type TextBlockProps = {
  text?: string;
  style?: string;
  marks?: string[];
};
export const textBlock = (props?: TextBlockProps) => {
  return getTestBlock({
    style: props?.style || "normal",

    children: [
      //@ts-ignore
      span({ text: props?.text || "textBlock", marks: props?.marks || [] }),
    ],
  });
};
