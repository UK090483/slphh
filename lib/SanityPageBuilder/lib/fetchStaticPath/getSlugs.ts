import { LocationConfig } from "@lib/SanityPageBuilder/types";
import type { SanityClient } from "@sanity/client/sanityClient";
import { defaultLocales } from "./defaultValues";

export type getSlugsProps = {
  doc: string;
  client: SanityClient;
  locales: LocationConfig;
  query?: string;
};
type slugResult = {
  slug: string;
  [k: string]: any;
};

export type GetSlugsResult = {
  slugs: slugResult[];
  homeSlug?: slugResult;
};

const getAllSlugs: (props: getSlugsProps) => Promise<GetSlugsResult> = async (
  props
) => {
  const { doc, client, locales = defaultLocales } = props;

  const i18nQuery = Object.entries(locales).reduce((acc, [locale, item]) => {
    if (!item.isDefault) {
      return `${acc} 'slug-${locale}': select(
        defined(pageType) => coalesce(pageType->slug_${locale}.current, pageType->slug.current) + '/' +  slug_${locale}.current,
        slug_${locale}.current
        ), `;
    }
    return acc;
  }, "");

  const allPages = await client.fetch<GetSlugsResult>(
    `{'slugs': *[_type == "${doc}"]{ 
          'slug': select(defined(pageType)=> pageType->slug.current +'/'+ slug.current,slug.current ),
          ${i18nQuery}
        },
        'homeSlug':{
            'slug':*[_id == 'siteConfig'][0].indexPage->slug.current
            }
        }`
  );

  if (!allPages || !allPages.slugs || !Array.isArray(allPages.slugs)) {
    throw new Error("No Path returned");
  }

  return allPages;
};

export default getAllSlugs;
