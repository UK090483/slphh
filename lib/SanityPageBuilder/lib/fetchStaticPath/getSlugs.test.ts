import getSlugs from "./getSlugs";
import { mockClient } from "../tests/testPrepare";

const locales = {
  de: { title: "Deutsch", isDefault: true, flag: "ðŸ‡©ðŸ‡ª" },
  en: { title: "Englisch", flag: "ðŸ‡ºðŸ‡¸" },
  da: { title: "DÃ¤nisch", flag: "ðŸ‡©ðŸ‡°" },
};

const database = [
  ...["testPageHome", "testPage", "widthSubRoute", "withString/subRoute"].map(
    (i) => ({
      _id: i,
      _type: "page",
      pageType: i === "widthSubRoute" ? { _ref: "testPageType" } : null,
      slug: { current: i },
      slug_en: { current: i + "en" },
      slug_da: { current: i + "da" },
    })
  ),
  {
    _id: "noSlug",
    _type: "page",
    slug_en: { current: "noSlugEn" },
    slug_da: { current: "noSlugEn" },
  },
  {
    _id: "noTransSlug",
    _type: "page",
    slug: { current: "noTransSlug" },
  },
  {
    _id: "testPageType",
    _type: "pageType",
    slug: { current: "testPageType" },
    slug_en: { current: "noSlugEn" },
  },
  { _id: "siteConfig", indexPage: { _ref: "testPageHome" } },
];

const mockFetchGetSlugs = ({
  fetchReturn,
  database,
  locales,
}: {
  fetchReturn?: any;
  database?: any;
  locales?: any;
}) => {
  return getSlugs({
    doc: "page",
    client: mockClient({
      ...(fetchReturn && { fetchReturn }),
      ...(database && { database }),
    }),
    locales: locales,
  });
};

describe("getSlugs", () => {
  it("should trow error if fetch result is not usable ", async () => {
    await expect(mockFetchGetSlugs({ fetchReturn: null })).rejects.toThrow();
    await expect(mockFetchGetSlugs({ fetchReturn: "st" })).rejects.toThrow();
  });

  it("should return right data no locales ", async () => {
    await expect(
      mockFetchGetSlugs({ fetchReturn: { slugs: [], homeSlug: null } })
    ).resolves.toStrictEqual({ slugs: [], homeSlug: null });
    await expect(mockFetchGetSlugs({ database })).resolves.toMatchSnapshot();
  });

  it("should return right data  width locales", async () => {
    await expect(
      mockFetchGetSlugs({ fetchReturn: { slugs: [], homeSlug: null }, locales })
    ).resolves.toStrictEqual({ slugs: [], homeSlug: null });

    await expect(
      mockFetchGetSlugs({ database, locales })
    ).resolves.toMatchSnapshot();
  });
});
