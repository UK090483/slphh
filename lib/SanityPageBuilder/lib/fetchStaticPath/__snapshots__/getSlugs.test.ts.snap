// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`getSlugs should return right data  width locales 1`] = `
Object {
  "homeSlug": Object {
    "slug": "testPageHome",
  },
  "slugs": Array [
    Object {
      "slug": "testPageHome",
      "slug-da": "testPageHomeda",
      "slug-en": "testPageHomeen",
    },
    Object {
      "slug": "testPage",
      "slug-da": "testPageda",
      "slug-en": "testPageen",
    },
    Object {
      "slug": "testPageType/widthSubRoute",
      "slug-da": "testPageType/widthSubRouteda",
      "slug-en": "noSlugEn/widthSubRouteen",
    },
    Object {
      "slug": "withString/subRoute",
      "slug-da": "withString/subRouteda",
      "slug-en": "withString/subRouteen",
    },
    Object {
      "slug": null,
      "slug-da": "noSlugEn",
      "slug-en": "noSlugEn",
    },
    Object {
      "slug": "noTransSlug",
      "slug-da": null,
      "slug-en": null,
    },
  ],
}
`;

exports[`getSlugs should return right data no locales  1`] = `
Object {
  "homeSlug": Object {
    "slug": "testPageHome",
  },
  "slugs": Array [
    Object {
      "slug": "testPageHome",
    },
    Object {
      "slug": "testPage",
    },
    Object {
      "slug": "testPageType/widthSubRoute",
    },
    Object {
      "slug": "withString/subRoute",
    },
    Object {
      "slug": null,
    },
    Object {
      "slug": "noTransSlug",
    },
  ],
}
`;
